src/IDNAConverterTrait.php

```diff
@@ -222,9 +224,19 @@ private function setLabels($domain = null): array
         if (!preg_match($pattern, $formatted_domain)) {
             throw new InvalidDomain(sprintf('The domain `%s` is invalid: the labels are malformed', $domain));
         }
+        
+        $ascii_domain = $this->idnToAscii($domain, $asciiOption);
 
-        $ascii_domain = $this->idnToAscii($domain);
-
-        return array_reverse(explode('.', $this->idnToUnicode($ascii_domain)));
+        return array_reverse(explode('.', $this->idnToUnicode($ascii_domain, $unicodeOption)));
+    }
+    
+    private function hasTransitionalDifference($domain):bool
+    {
+        if(!is_string($domain) || empty($domain)){
+            return false;
+        }
+        $info = [];
+        idn_to_ascii($domain, IDNA_NONTRANSITIONAL_TO_ASCII, INTL_IDNA_VARIANT_UTS46, $info);
```

Comment:
I'm wondering if we should not refactor this line so that `idn_to_ascii` is only use once with all the checks in the trait 🤔 .  so idnToAscii and hasTransitonalDifference uses the same method ? What do you think this way maintenance becomes easier in case of bugfix or improvement ?

Reply:
this method is not so heavy and not raise additional exceptions; In previous variant we use it once and save state in private property, but it  provide additional problems https://github.com/jeremykendall/php-domain-parser/pull/235#discussion_r273705257

Reply:
I was thinking of something along these lines:

```php
    private function idnToAscii(string $domain, int $options): string
    {
    	list($domain, $isTransitionalDifferent) = $this->transformToAscii($domain, $options);

    	return $domain;
    }

	private function hasTransitionalDifference($domain):bool
	{
	        if (!is_string($domain) || empty($domain)){
	            return false;
	        }

	        list($domain, $isTransitionalDifferent) = $this->transformToAscii($domain, $options);

	        return $isTransitionalDifferent;
	    }
	}

	private function transformToAscii(string $domain, int $options): array
	{
        $domain = rawurldecode($domain);
        static $pattern = '/[^\x20-\x7f]/';
        if (!preg_match($pattern, $domain)) {
            return [strtolower($domain), false];
        }
        $output = idn_to_ascii($domain, 0, INTL_IDNA_VARIANT_UTS46, $arr);
        if (0 !== $arr['errors']) {
            throw new InvalidDomain(sprintf('The host `%s` is invalid : %s', $domain, self::getIdnErrors($arr['errors'])));
        }
        // @codeCoverageIgnoreStart
        if (false === $output) {
            throw new UnexpectedValueException(sprintf('The Intl extension is misconfigured for %s, please correct this issue before proceeding.', PHP_OS));
        }
        // @codeCoverageIgnoreEnd
        if (false === strpos($output, '%')) {
        	return [$output, $result['isTransitionalDifferent']];
        }
        
        throw new InvalidDomain(sprintf('The host `%s` is invalid: it contains invalid characters', $domain));	
	}
```

Reply:
Also could you update your code to **NOT** use `empty` as I tend to remove its usage from the codebase. I generally prefer `[] === $array`  or clearly stating what domain is. Empty leads to error because the empty string and '0' are all considered empty :'(